File: experiment-3/runs/language=go_model=claude-opus-4-7_tooling=none/rep1/protocol.go
// Minimal MCP (Model Context Protocol) implementation: JSON-RPC 2.0 over
// stdio, supporting initialize, tools/list, tools/call and ping.
package main

import "encoding/json"

const (
	protocolVersion = "2025-06-18"
	serverName      = "brazilian-soccer-mcp"
	serverVersion   = "1.0.0"
)

type rpcRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

type rpcResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Result  any             `json:"result,omitempty"`
	Error   *rpcError       `json:"error,omitempty"`
}

type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// isNotification reports whether the request carries no id (and so expects no
// response).
func (r *rpcRequest) isNotification() bool { return len(r.ID) == 0 }

// handleRequest processes a single JSON-RPC request and returns the response,
// or nil when the request is a notification.
func handleRequest(db *DB, req *rpcRequest) *rpcResponse {
	resp := &rpcResponse{JSONRPC: "2.0", ID: req.ID}
	switch req.Method {
	case "initialize":
		resp.Result = initializeResult(req.Params)
	case "notifications/initialized", "notifications/cancelled":
		return nil
	case "ping":
		resp.Result = map[string]any{}
	case "tools/list":
		resp.Result = map[string]any{"tools": toolSchemas()}
	case "tools/call":
		resp.Result = handleToolCall(db, req.Params)
	default:
		if req.isNotification() {
			return nil
		}
		resp.Error = &rpcError{Code: -32601, Message: "method not found: " + req.Method}
	}
	if req.isNotification() {
		return nil
	}
	return resp
}

func initializeResult(params json.RawMessage) map[string]any {
	version := protocolVersion
	if len(params) > 0 {
		var p struct {
			ProtocolVersion string `json:"protocolVersion"`
		}
		if json.Unmarshal(params, &p) == nil && p.ProtocolVersion != "" {
			version = p.ProtocolVersion
		}
	}
	return map[string]any{
		"protocolVersion": version,
		"capabilities":    map[string]any{"tools": map[string]any{}},
		"serverInfo":      map[string]any{"name": serverName, "version": serverVersion},
	}
}

// handleToolCall dispatches a tools/call request to the named tool and wraps
// the result in MCP content format.
func handleToolCall(db *DB, params json.RawMessage) map[string]any {
	var p struct {
		Name      string         `json:"name"`
		Arguments map[string]any `json:"arguments"`
	}
	if err := json.Unmarshal(params, &p); err != nil {
		return toolErrorResult("invalid tools/call parameters: " + err.Error())
	}
	tool, ok := toolByName[p.Name]
	if !ok {
		return toolErrorResult("unknown tool: " + p.Name)
	}
	text, err := tool.Handler(db, p.Arguments)
	if err != nil {
		return toolErrorResult(err.Error())
	}
	return map[string]any{
		"content": []map[string]any{{"type": "text", "text": text}},
	}
}

func toolErrorResult(msg string) map[string]any {
	return map[string]any{
		"content": []map[string]any{{"type": "text", "text": msg}},
		"isError": true,
	}
}

File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=TDD/rep1-failed/lib/brasil_soccer/mcp/protocol.ex
defmodule BrasilSoccer.MCP.Protocol do
  @moduledoc """
  Pure JSON-RPC 2.0 dispatch for the Model Context Protocol.

  `handle/2` takes a decoded request map and the in-memory dataset and returns a
  response map to encode and send back, or `nil` for notifications (which must
  not be answered). Keeping this pure makes the protocol behaviour fully
  unit-testable without any IO.
  """

  alias BrasilSoccer.MCP.Tools

  @protocol_version "2024-11-05"
  @server_name "brazil-soccer-mcp"
  @server_version "0.1.0"

  @doc "Handle a single decoded JSON-RPC request against `data`."
  @spec handle(map(), Tools.data()) :: map() | nil
  def handle(%{"method" => method} = req, data) do
    id = req["id"]
    params = req["params"] || %{}

    cond do
      notification?(method) -> nil
      true -> dispatch(method, params, id, data)
    end
  end

  defp notification?("notifications/" <> _), do: true
  defp notification?(_), do: false

  defp dispatch("initialize", _params, id, _data) do
    result(id, %{
      "protocolVersion" => @protocol_version,
      "capabilities" => %{"tools" => %{"listChanged" => false}},
      "serverInfo" => %{"name" => @server_name, "version" => @server_version}
    })
  end

  defp dispatch("ping", _params, id, _data), do: result(id, %{})

  defp dispatch("tools/list", _params, id, _data) do
    result(id, %{"tools" => Tools.specs()})
  end

  defp dispatch("tools/call", params, id, data) do
    name = params["name"]
    args = params["arguments"] || %{}

    case Tools.call(name, args, data) do
      {:ok, text} -> result(id, tool_content(text, false))
      {:error, message} -> result(id, tool_content(message, true))
    end
  end

  defp dispatch(method, _params, id, _data) do
    error(id, -32601, "Method not found: #{method}")
  end

  defp tool_content(text, is_error?) do
    %{"content" => [%{"type" => "text", "text" => text}], "isError" => is_error?}
  end

  defp result(id, result) do
    %{"jsonrpc" => "2.0", "id" => id, "result" => result}
  end

  defp error(id, code, message) do
    %{"jsonrpc" => "2.0", "id" => id, "error" => %{"code" => code, "message" => message}}
  end
end
